test/orange/orangeinp/CsgTreeUtils.test.cc

```diff
@@ -385,6 +385,306 @@ TEST_F(CsgTreeUtilsTest, calc_surfaces)
+        "all{5,6,9}, }",
+        to_string(simplified));
+
+    // Check a well-formed tree
+    this->insert(Negated{j3});
+    EXPECT_EQ(
+        "{0: true, 1: not{0}, 2: surface 0, 3: surface 1, 4: not{3}, 5: "
+        "all{2,4}, 6: not{5}, 7: any{2,4}, 8: not{7}, 9: surface 2, 10: "
+        "not{9}, 11: all{5,7}, 12: not{11}, 13: all{6,8,12}, 14: not{13}, }",
+        to_string(tree_));
+
+    // Complex case with a negated join with negated join as children
+    simplified = transform_negated_joins(tree_);
+    EXPECT_EQ(
+        "{0: true, 1: not{0}, 2: surface 0, 3: not{2}, 4: surface 1, 5: "
+        "not{4}, 6: any{3,4}, 7: all{2,5}, 8: all{3,4}, 9: any{2,5}, 10: "
+        "surface 2, 11: not{10}, 12: any{6,8}, 13: all{7,9}, 14: any{7,9,13}, "
+        "}",
+        to_string(simplified));
+
+    tree_ = {};
+
+    s0 = this->insert(S{0});
+    s1 = this->insert(S{1});
+    n0 = this->insert(Negated{s0});
+    n1 = this->insert(Negated{s1});
+    j0 = this->insert(Joined{op_or, {n0, n1}});
+    n3 = this->insert(Negated{j0});
+    s2 = this->insert(S{2});
+    auto n4 = this->insert(Negated{s2});
+    this->insert(Joined{op_and, {n3, n4}});
+    // Check a well-formed tree
+    EXPECT_EQ(
+        "{0: true, 1: not{0}, 2: surface 0, 3: surface 1, 4: not{2}, 5: "
+        "not{3}, 6: any{4,5}, 7: not{6}, 8: surface 2, 9: not{8}, 10: "
+        "all{7,9}, }",
+        to_string(tree_));
+
+    // Complex case with a negated join with negated children
+    simplified = transform_negated_joins(tree_);
+    EXPECT_EQ(
+        "{0: true, 1: not{0}, 2: surface 0, 3: surface 1, 4: all{2,3}, 5: "
+        "surface 2, 6: not{5}, 7: all{4,6}, }",
+        to_string(simplified));
+}
+
+TEST_F(CsgTreeUtilsTest, transform_negated_joins_with_volumes)
+{
+    auto s0 = this->insert(S{0});
+    auto s1 = this->insert(S{1});
+    auto n0 = this->insert(Negated{s0});
+    auto n1 = this->insert(Negated{s1});
+    auto j0 = this->insert(Joined{op_or, {n0, n1}});
+    auto n3 = this->insert(Negated{j0});
+    auto s2 = this->insert(S{2});
+    auto n4 = this->insert(Negated{s2});
+    auto j1 = this->insert(Joined{op_and, {n3, n4}});
+    tree_.insert_volume(j0);
+    tree_.insert_volume(j1);
+    tree_.insert_volume(n3);
+    // Check a well-formed tree
+    EXPECT_EQ(
+        "{0: true, 1: not{0}, 2: surface 0, 3: surface 1, 4: not{2}, 5: "
+        "not{3}, 6: any{4,5}, 7: not{6}, 8: surface 2, 9: not{8}, 10: "
+        "all{7,9}, }",
+        to_string(tree_));
+
+    // Complex case with a negated join with negated children
+    auto simplified = transform_negated_joins(tree_);
+    EXPECT_EQ(
+        "{0: true, 1: not{0}, 2: surface 0, 3: surface 1, 4: not{2}, 5: "
+        "not{3}, 6: all{2,3}, 7: any{4,5}, 8: surface 2, 9: not{8}, 10: "
+        "all{6,9}, }",
+        to_string(simplified));
+    // Check that the new volumes map to the correct node
+    auto& volumes = simplified.volumes();
+    EXPECT_EQ(volumes.size(), 3);
```

Comment:
Good to use this form to prevent the code crashing in case of an error...
```suggestion
    ASSERT_EQ(volumes.size(), 3);
```